### DISCUSSION TITLE
Incorrect error CS8602: Dereference of a possibly null reference on a guarded string value

### DISCUSSION THREAD

**Discussion by jozefizso** [dotnet/roslyn#73946] [Q&A]
**Version Used**: C# 8.0

**Steps to Reproduce**:

1. Create project targeting .NET Framework 4.8 and .NET 8, enable C# 8 and treat nullable warnings as compilation errors
2. Selectively enable nullables for a portion of a codebase
3. A nullable string variable is guarded in a static method, but the code cannot be compiled because accessing the variable after the guard will cause the CS8602 error.

A minimal repro: https://gist.github.com/jozefizso/ff7e0ce119366b1129cb6a0d5fb0edee

**Diagnostic Id**: CS8602: Dereference of a possibly null reference

**Expected Behavior**: The code should compile.

**Actual Behavior**: Compiler fails on `s.Length` expression while the `s` variable is reported as not null.

**Comment by CyrusNajmabadi** [comment]
.Net framework likely doesn't have annotations for string.IsNullOrEmpty.  I'd recommend you just do nullable testing for your .net8 target.

**Comment by huoyaoyuan** [comment]
See also #73823
